source/opt/loop_utils.cpp

```diff
@@ -0,0 +1,375 @@
+// Copyright (c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "opt/cfg.h"
+#include "opt/ir_builder.h"
+#include "opt/ir_context.h"
+#include "opt/loop_descriptor.h"
+#include "opt/loop_utils.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+// Return true if |bb| is dominated by at least one block in |exits|
+static inline bool DominatesAnExit(
+    ir::BasicBlock* bb, const std::unordered_set<ir::BasicBlock*>& exits,
+    const opt::DominatorTree& dom_tree) {
+  for (ir::BasicBlock* e_bb : exits)
+    if (dom_tree.Dominates(bb, e_bb)) return true;
+  return false;
+}
+
+// Utility class to rewrite uses in terms of phi nodes to achieve a LCSSA form.
+// It works by inserting SSA nodes where needed to make an out-of-loop use of a
+// def dependent on the exiting phi node.
+class LCSSARewriter {
+ public:
+  LCSSARewriter(ir::IRContext* context, const opt::DominatorTree& dom_tree,
+                const std::unordered_set<ir::BasicBlock*>& exit_bb,
+                const ir::Instruction& def_insn)
+      : context_(context),
+        cfg_(context_->cfg()),
+        dom_tree_(dom_tree),
+        insn_type_(def_insn.type_id()),
+        exit_bb_(exit_bb) {}
+
+  // Rewrites the use of |def_insn_| by the instruction |user| at the index
+  // |operand_index| in terms of phi instruction.
+  // This recursively builds new phi instructions from |user| to the loop exit
+  // blocks' phis.
+  // The use of |def_insn_| in |user| is replaced by the relevant phi
+  // instruction at the end of the operation.
+  // This operation does not update the def/use manager, instead it records what
+  // needs to be updated. The actual update is performed by UpdateManagers.
```

Comment:
I think clang-format has messed up these comments. If you make the whole block a single line and run clang format again it should fix it